File: apps/server/services/ai_router/telemetry.py
"""Request/attempt telemetry persistence for the RADAS 9Router module.

Log rows carry metadata only — request IDs, provider/model resolution, error
classifications, token counts, and cost estimates. Prompts, completions, and
credentials are never persisted here; redaction is asserted by tests.
"""
from __future__ import annotations

import json
import logging
import time
import uuid
from typing import Any, Dict, List, Optional

from storage import pg

logger = logging.getLogger(__name__)


def record_request_log(
    *,
    org_id: str,
    user_id: str,
    endpoint: str,
    requested_model: str,
    attempts: List[Dict[str, Any]],
    status: str,
    request_id: str,
    resolved_provider: str = "",
    resolved_model: str = "",
    error_code: str = "",
    http_status: Optional[int] = None,
    latency_ms: int = 0,
    prompt_tokens: int = 0,
    completion_tokens: int = 0,
    tokens_saved_rtk: int = 0,
    stream: bool = False,
    created_at: Optional[float] = None,
) -> None:
    """Persist one redacted request log row; telemetry must never fail a request."""
    fallback_used = sum(1 for attempt in attempts if attempt.get("status") == "error") > 0
    try:
        from .pricing import estimate_cost

        cost = estimate_cost(resolved_model or requested_model, prompt_tokens, completion_tokens)
        pg.execute(
            """INSERT INTO org_ai_request_logs
               (id, org_id, user_id, endpoint, requested_model, resolved_provider, resolved_model,
                status, error_code, http_status, latency_ms, prompt_tokens, completion_tokens,
                tokens_saved_rtk, cost_usd_est, fallback_used, stream, request_id, attempts, created_at)
               VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)""",
            (
                f"log-{uuid.uuid4().hex[:12]}",
                org_id,
                user_id or "system",
                endpoint,
                requested_model,
                resolved_provider or None,
                resolved_model or None,
                status,
                error_code or None,
                http_status,
                max(0, int(latency_ms)),
                max(0, int(prompt_tokens)),
                max(0, int(completion_tokens)),
                max(0, int(tokens_saved_rtk)),
                cost,
                fallback_used,
                stream,
                request_id,
                json.dumps(attempts),
                created_at if created_at is not None else time.time(),
            ),
        )
    except Exception:
        logger.warning("9Router request log write failed for request %s", request_id, exc_info=True)


def list_request_logs(
    org_id: str,
    *,
    limit: int = 50,
    since: Optional[float] = None,
    until: Optional[float] = None,
    status: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """Redacted log rows for one organization, newest first."""
    clauses = ["org_id = %s"]
    params: List[Any] = [org_id]
    if since is not None:
        clauses.append("created_at >= %s")
        params.append(float(since))
    if until is not None:
        clauses.append("created_at <= %s")
        params.append(float(until))
    if status:
        clauses.append("status = %s")
        params.append(status)
    params.append(max(1, min(200, int(limit))))
    rows = pg.query_all(
        f"SELECT * FROM org_ai_request_logs WHERE {' AND '.join(clauses)} ORDER BY created_at DESC LIMIT %s",
        tuple(params),
    )
    for row in rows:
        if isinstance(row.get("attempts"), str):
            try:
                row["attempts"] = json.loads(row["attempts"])
            except (TypeError, ValueError):
                row["attempts"] = []
    return rows


def cost_summary(org_id: str, *, since: Optional[float] = None, until: Optional[float] = None) -> Dict[str, Any]:
    """Aggregate cost/token/latency estimates over a date range (estimates only)."""
    clauses = ["org_id = %s", "status = 'success'"]
    params: List[Any] = [org_id]
    if since is not None:
        clauses.append("created_at >= %s")
        params.append(float(since))
    if until is not None:
        clauses.append("created_at <= %s")
        params.append(float(until))
    rows = pg.query_all(
        f"""SELECT resolved_provider, resolved_model,
                   COUNT(*) AS requests,
                   SUM(prompt_tokens) AS prompt_tokens,
                   SUM(completion_tokens) AS completion_tokens,
                   SUM(tokens_saved_rtk) AS tokens_saved_rtk,
                   SUM(cost_usd_est) AS cost_usd_est,
                   AVG(latency_ms) AS avg_latency_ms,
                   SUM(CASE WHEN fallback_used THEN 1 ELSE 0 END) AS fallbacks
            FROM org_ai_request_logs WHERE {' AND '.join(clauses)}
            GROUP BY resolved_provider, resolved_model
            ORDER BY cost_usd_est DESC""",
        tuple(params),
    )
    breakdown = [
        {
            "provider": row.get("resolved_provider") or "unknown",
            "model": row.get("resolved_model") or "unknown",
            "requests": int(row.get("requests") or 0),
            "prompt_tokens": int(row.get("prompt_tokens") or 0),
            "completion_tokens": int(row.get("completion_tokens") or 0),
            "tokens_saved_rtk": int(row.get("tokens_saved_rtk") or 0),
            "cost_usd_est": round(float(row.get("cost_usd_est") or 0), 6),
            "avg_latency_ms": round(float(row.get("avg_latency_ms") or 0), 1),
            "fallbacks": int(row.get("fallbacks") or 0),
        }
        for row in rows
    ]
    return {
        "total_requests": sum(entry["requests"] for entry in breakdown),
        "total_prompt_tokens": sum(entry["prompt_tokens"] for entry in breakdown),
        "total_completion_tokens": sum(entry["completion_tokens"] for entry in breakdown),
        "total_tokens_saved_rtk": sum(entry["tokens_saved_rtk"] for entry in breakdown),
        "total_fallbacks": sum(entry["fallbacks"] for entry in breakdown),
        "total_cost_usd_est": round(sum(entry["cost_usd_est"] for entry in breakdown), 6),
        "breakdown": breakdown,
        "note": "Costs are public-rate estimates for observability, not billing data.",
    }

File: apps/server/lib/radas_ai/telemetry.ex
defmodule RadasAI.Telemetry do
  @moduledoc """
  Port of `services/ai_router/telemetry.py`.

  Request/attempt telemetry persistence. Log rows carry metadata only —
  request IDs, provider/model resolution, error classifications, token counts,
  and cost estimates. Prompts, completions, and credentials are never
  persisted; telemetry must never fail a request.
  """

  import RadasAI.DB

  alias RadasAI.Pricing

  @doc "Persist one redacted request log row (fail-open)."
  @spec record_request_log(keyword()) :: :ok
  def record_request_log(opts) do
    org_id = Keyword.fetch!(opts, :org_id)
    user_id = Keyword.get(opts, :user_id, "")
    endpoint = Keyword.fetch!(opts, :endpoint)
    requested_model = Keyword.fetch!(opts, :requested_model)
    attempts = Keyword.get(opts, :attempts, [])
    status = Keyword.fetch!(opts, :status)
    request_id = Keyword.fetch!(opts, :request_id)

    fallback_used =
      Enum.any?(attempts, &(is_map(&1) and Map.get(&1, "status") == "error"))

    cost =
      try do
        Pricing.estimate_cost(Keyword.get(opts, :resolved_model) || requested_model, Keyword.get(opts, :prompt_tokens, 0), Keyword.get(opts, :completion_tokens, 0))
      rescue
        _ -> 0.0
      end

    execute!(
      """
      INSERT INTO org_ai_request_logs
        (id, org_id, user_id, endpoint, requested_model, resolved_provider, resolved_model,
         status, error_code, http_status, latency_ms, prompt_tokens, completion_tokens,
         tokens_saved_rtk, cost_usd_est, fallback_used, stream, request_id, attempts, created_at)
      VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20)
      """,
      [
        "log-" <> uuid12(),
        org_id,
        present_or(user_id, "system"),
        endpoint,
        requested_model,
        Keyword.get(opts, :resolved_provider),
        Keyword.get(opts, :resolved_model),
        status,
        present_or(Keyword.get(opts, :error_code), nil),
        Keyword.get(opts, :http_status),
        max(0, Keyword.get(opts, :latency_ms, 0) |> trunc()),
        max(0, Keyword.get(opts, :prompt_tokens, 0)),
        max(0, Keyword.get(opts, :completion_tokens, 0)),
        max(0, Keyword.get(opts, :tokens_saved_rtk, 0)),
        cost,
        fallback_used,
        Keyword.get(opts, :stream, false),
        request_id,
        Jason.encode!(Enum.map(attempts, &Map.new/1)),
        Keyword.get(opts, :created_at) || now()
      ]
    )

    :ok
  rescue
    _ -> :ok
  end

  @doc "Redacted log rows for one organization, newest first."
  @spec list_request_logs(String.t(), keyword()) :: [map()]
  def list_request_logs(org_id, opts \\ []) do
    limit = clamp(Keyword.get(opts, :limit, 50), 1, 200)
    since = Keyword.get(opts, :since)
    until = Keyword.get(opts, :until)
    status = Keyword.get(opts, :status)

    {clauses, params} =
      build_filters([
        {"org_id = $", org_id},
        {"created_at >= $", since},
        {"created_at <= $", until},
        {"status = $", status}
      ])

    sql =
      "SELECT * FROM org_ai_request_logs WHERE #{clauses} ORDER BY created_at DESC LIMIT $" <>
        Integer.to_string(length(params) + 1)

    # psycopg auto-decodes JSONB; mirror that for the attempts column.
    Enum.map(query_all!(sql, params ++ [limit]), fn row ->
      case row["attempts"] do
        binary when is_binary(binary) ->
          case Jason.decode(binary) do
            {:ok, list} when is_list(list) -> Map.put(row, "attempts", list)
            _ -> row
          end

        _ ->
          row
      end
    end)
  end

  @doc "Aggregate cost/token/latency estimates over a date range (estimates only)."
  @spec cost_summary(String.t(), keyword()) :: map()
  def cost_summary(org_id, opts \\ []) do
    since = Keyword.get(opts, :since)
    until = Keyword.get(opts, :until)

    {clauses, params} =
      build_filters([
        {"org_id = $", org_id},
        {"status = 'success'", nil},
        {"created_at >= $", since},
        {"created_at <= $", until}
      ])

    rows =
      query_all!(
        """
        SELECT resolved_provider, resolved_model,
               COUNT(*) AS requests,
               SUM(prompt_tokens) AS prompt_tokens,
               SUM(completion_tokens) AS completion_tokens,
               SUM(tokens_saved_rtk) AS tokens_saved_rtk,
               SUM(cost_usd_est) AS cost_usd_est,
               AVG(latency_ms) AS avg_latency_ms,
               SUM(CASE WHEN fallback_used THEN 1 ELSE 0 END) AS fallbacks
        FROM org_ai_request_logs WHERE #{clauses}
        GROUP BY resolved_provider, resolved_model
        ORDER BY cost_usd_est DESC
        """,
        params
      )

    breakdown =
      Enum.map(rows, fn row ->
        %{
          "provider" => row["resolved_provider"] || "unknown",
          "model" => row["resolved_model"] || "unknown",
          "requests" => row["requests"] || 0,
          "prompt_tokens" => row["prompt_tokens"] || 0,
          "completion_tokens" => row["completion_tokens"] || 0,
          "tokens_saved_rtk" => row["tokens_saved_rtk"] || 0,
          "cost_usd_est" => round6(row["cost_usd_est"]),
          "avg_latency_ms" => round1(row["avg_latency_ms"]),
          "fallbacks" => row["fallbacks"] || 0
        }
      end)

    %{
      "total_requests" => sum_by(breakdown, "requests"),
      "total_prompt_tokens" => sum_by(breakdown, "prompt_tokens"),
      "total_completion_tokens" => sum_by(breakdown, "completion_tokens"),
      "total_tokens_saved_rtk" => sum_by(breakdown, "tokens_saved_rtk"),
      "total_fallbacks" => sum_by(breakdown, "fallbacks"),
      "total_cost_usd_est" => breakdown |> Enum.map(& &1["cost_usd_est"]) |> Enum.sum() |> round6(),
      "breakdown" => breakdown,
      "note" => "Costs are public-rate estimates for observability, not billing data."
    }
  end

  # -- helpers ---------------------------------------------------------------

  defp build_filters(entries) do
    entries
    |> Enum.reject(fn {_clause, value} -> is_nil(value) end)
    |> Enum.reduce({[], []}, fn
      {clause, value}, {cls, ps} = _acc ->
        if String.ends_with?(clause, "$") do
          {cls ++ ["(#{clause}#{length(ps) + 1})"], ps ++ [value]}
        else
          {cls ++ ["(#{clause})"], ps}
        end
    end)
    |> then(fn {cls, ps} -> {Enum.join(cls, " AND "), ps} end)
  end

  defp present_or(value, fallback) do
    if value in [nil, ""], do: fallback, else: value
  end

  defp clamp(value, min, max), do: value |> max(min) |> min(max)

  defp sum_by(breakdown, key), do: breakdown |> Enum.map(& &1[key]) |> Enum.sum()

  defp round6(nil), do: 0.0
  defp round6(value), do: Float.round(to_float(value), 6)

  defp round1(nil), do: 0.0
  defp round1(value), do: Float.round(to_float(value), 1)

  # PostgreSQL aggregates (AVG/SUM on DOUBLE PRECISION) arrive as Decimal.
  defp to_float(%Decimal{} = value), do: Decimal.to_float(value)
  defp to_float(value) when is_float(value), do: value
  defp to_float(value) when is_integer(value), do: value * 1.0
  defp to_float(value) when is_binary(value) do
    case Float.parse(value) do
      {f, _} -> f
      :error -> 0.0
    end
  end

  defp uuid12 do
    :crypto.strong_rand_bytes(6) |> Base.encode16(case: :lower)
  end
end
